### PR TITLE
Alternative build method

### PR DESCRIPTION
**Pull request by mauricioszabo**
I decided to draft this PR to request for comments (I think "code' speaks better than words in this case). So, this PR fundamentally changes how we create Atom binaries.

## Motivation

Currently, Atom builds binaries with a combination of `./script/bootstrap` and `./script/build`. I could not run `./script/boostrap` in a reliable way, and discussing with some people it seems that it does not function with recent `npm` versions. Also, all the scripts to build Atom end up adding more than 5.000 lines of code that the community will have to both understand and maintain, so I decided to make a quick spike to see how hard it would be to build Atom without these scripts

## How is being done?

I'm using `yarn` instead of `npm` and I'm adding `electron-builder`, Electron and some other packages to `dev-dependencies` for easier bootrapping of the app. I also updated a doc to show how to start developing Atom with this method.

## What is not working?

Unfortunately, this method is not 100% complete yet. Atom works, but `github` package is developed with ESM modules, so it can't be loaded as of now (probably the boostrap script ends up using `babel` to transpile code). Other packages work, but another problem is that "core" packages are missing their README files for some reason. So the list of things I know we'll have to take a look:

1. [ ] Github package is not working
2. [ ] README of "core" packages is also not working
3. [ ] When we start Atom, for some reason it locks the terminal - as if we're always running Atom with the `-f` switch
4. [ ] V8 Snapshots doesn't currently work - seems that they also need Babel to transpile packages
5. [ ] `apm` does not get installed on the PATH - this may need to be fixed

## What is the advantage?

Updating some "core" packages like Electron and Tree-Sitter currently is very hard. Atom solution is to "boostrap" and then point an existing Atom editor to the "bootstrapped" package, which is not that reliable and also does not help on the Electron case. Also, it's all made with custom scripts, which are not reliable as they are now. We may fix the scripts to work 100% of the time, but IMHO it's better to embrace what the Node/JS/Electron community is using so we gain bugfixes and access to more targets when they appear.

Also, the current package is able to auto-update an `AppImage` on Linux if we decide to implement this (`electron-builder` have a way to implement a cross-platform independent auto-updater). It is also able to draft releases on GitHub too, and the number of lines of code that I had to write to make it work is really small compared with the scripts directory.

Should we decide to go through this route, we have to decide if we're going to keep the idea of using "V8 snapshots" to load Atom faster, and if we want to transpile code. Also, maybe it's a good idea to, in the future, change the way Atom detects "core" packages - currently, it's somewhat weird and depends on adding stuff on `package.json`. Maybe a better way could be to just make Atom search for packages in two directories - the `ATOM_HOME` and the "asar" file when bundled.

### CODE REVIEW

**Comment by mauricioszabo** [comment]
Forgot to mention: I want comments if people think this is a move in the right direction. If you think so, I'll work to fix the issues I mentioned on this PR :)

**Review by aminya — CHANGES_REQUESTED** [review]
I like the approach to modernizing the build script. 👍🏻 

However, I am not sure if this covers all the things that the current build script does. So it needs some close comparison and verification.

Some things don't match how we have been working in the organization. For example, we have been trying to use `pnpm` instead of `apm` for years. So, this should change `yarn` back to `apm/npm`  (since the pnpm conversion is happening on other branches).


Once this is ready, we should probably remove the old method. We don't want to maintain two parallel ways to build Atom.

**Comment by mauricioszabo** [comment]
> I like the approach to modernizing the build script. 👍🏻

Good, so I'll keep working on it. What I know is that it currently is not making any transpilation and it's not generating the V8 snapshots, so I'll try to make these work and update this PR :)

**Comment by mauricioszabo** [comment]
> Some things don't match how we have been working in the organization. For example, we have been trying to use pnpm instead of apm for years. So, this should change yarn back to apm/npm (since the pnpm conversion is happening on other branches).

Honestly, I don't want to keep using yarn - if possible, I would like to mibrate to `pnpm` but that's what it is currently working. I'll revert back to `pnpm` (or even `npm`) when that works reliably, the choice for `yarn` was because it literally was the only thing that worked 100% of the time...

**Comment by aminya** [comment]
I changed the priority to low as it is not urgent.

**Comment by DeeDeeG** [comment]
`github` package should work if transpiled properly.

That is currently handled by [script/lib/transpile-packages-with-custom-transpiler-paths.js](https://github.com/atom-community/atom/blob/master/script/lib/transpile-packages-with-custom-transpiler-paths.js)

See also upstream's version of the same file: https://github.com/atom/atom/blob/master/script/lib/transpile-packages-with-custom-transpiler-paths.js)

File: package.json [hunk 1]
```diff
@@ -284,5 +287,10 @@
       "atom",
       "snapshotResult"
     ]
+  },
+  "devDependencies": {
+    "electron": "11.5.0",
+    "electron-builder": "^23.1.0",
+    "electron-rebuild": "^3.2.7"
   }
```
Comment (aminya) [inline package.json:295]: These are going to break the snapshot script. They should be moved to `script/package.json`

File: package.json [hunk 2]
```diff
@@ -270,7 +269,11 @@
   "private": true,
   "scripts": {
     "preinstall": "node -e 'process.exit(0)'",
-    "test": "node script/test"
+    "test": "node script/test",
+    "build": "electron-rebuild",
+    "build:apm": "cd apm; yarn install; ./node_modules/atom-package-manager/bin/npm rebuild",
+    "start": "electron .",
+    "dist": "node script/electron-builder.js"
```
Comment (aminya) [inline package.json:276]: I'd rather see the electron-builder as a part of the current build process. Then, the duplicate parts of the current build can be removed. See #311 for an example to run this. I did it for Parcel before generating the snapshots.

**Review by aminya — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (mauricioszabo) [reply package.json:276]: I actually made this experiment to see if we could _replace_ the current build process, as it seems completely unreliable. But I also see your point. Maybe we can discuss this here? https://github.com/atom-community/atom/discussions/443

Comment (aminya) [reply package.json:276]: I added my response here:
https://github.com/atom-community/atom/discussions/443#discussioncomment-3111302

File: script/electron-builder.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
+const path = require('path')
+const normalizePackageData = require('normalize-package-data');
+const fs = require("fs/promises");
```
Comment (aminya) [inline script/electron-builder.js:3]: `fs/promises` is only available on newer Node versions.

File: script/electron-builder.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    fs.mkdir('binaries').catch(() => "")
+    Promise.all(result.map(r => fs.copyFile(r, path.join('binaries', path.basename(r)))))
+  }).catch((error) => {
+    console.error("Error building binaries")
+    console.error(error)
+    process.exit(1)
+  })
+}
+
+main()
```
Comment (aminya) [inline script/electron-builder.js:104]: All the promises should be handled.

```suggestion
main().catch((err) => {
  throw err;
}
```

File: script/electron-builder.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    config: options
+  }).then((result) => {
+    console.log("Built binaries")
+    fs.mkdir('binaries').catch(() => "")
+    Promise.all(result.map(r => fs.copyFile(r, path.join('binaries', path.basename(r)))))
+  }).catch((error) => {
+    console.error("Error building binaries")
+    console.error(error)
+    process.exit(1)
+  })
```
Comment (aminya) [inline script/electron-builder.js:101]: Please use `async/await` and `try/catch`

File: script/electron-builder.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+async function main() {
+  const package = await fs.readFile('package.json', "utf-8")
+  options.extraMetadata = generateMetadata(JSON.parse(package))
+  builder.build({
+    config: options
+  }).then((result) => {
+    console.log("Built binaries")
+    fs.mkdir('binaries').catch(() => "")
```
Comment (aminya) [inline script/electron-builder.js:95]: Why do you ignore the error?

File: script/electron-builder.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    Promise.all(result.map(r => fs.copyFile(r, path.join('binaries', path.basename(r)))))
```
Comment (aminya) [inline script/electron-builder.js:96]: All the promises should be handled.

```suggestion
    await Promise.all(result.map(r => fs.copyFile(r, path.join('binaries', path.basename(r)))))
```

File: script/electron-builder.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      { "target": "nsis" },
+      { "target": "portable" }
+    ]
+  },
+  "extraMetadata": {
+  }
+}
+
+async function main() {
+  const package = await fs.readFile('package.json', "utf-8")
```
Comment (aminya) [inline script/electron-builder.js:89]: Please use the `CONFIG` object to get the absolute root path of the repository.

File: script/generate-metadata-for-builder.js [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+const fs = require('fs-plus');
+const normalizePackageData = require('normalize-package-data');
+const path = require('path');
+const semver = require('semver');
+
+module.exports = function(packageJSON) {
+  return {
+    _atomPackages: buildBundledPackagesMetadata(packageJSON),
+    _atomMenu: buildPlatformMenuMetadata(packageJSON),
+    _atomKeymaps: buildPlatformKeymapsMetadata(packageJSON)
```
Comment (aminya) [inline script/generate-metadata-for-builder.js:13]: Why do the properties have `_` prefixes? This is usually indicative of unused properties.

File: script/generate-metadata-for-builder.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
+'use strict';
+
+const CSON = require('season');
+const fs = require('fs-plus');
+const normalizePackageData = require('normalize-package-data');
+const path = require('path');
+const semver = require('semver');
+
+module.exports = function(packageJSON) {
```
Comment (aminya) [inline script/generate-metadata-for-builder.js:9]:
```suggestion
module.exports = function generateMetaDataForBuilder(packageJSON) {
```

I'd also rather not use "default" exports.

File: script/generate-metadata-for-builder.js [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    _atomPackages: buildBundledPackagesMetadata(packageJSON),
+    _atomMenu: buildPlatformMenuMetadata(packageJSON),
+    _atomKeymaps: buildPlatformKeymapsMetadata(packageJSON)
+  }
+};
+
+function buildBundledPackagesMetadata(packageJSON) {
+  const packages = {};
+  for (let packageName of Object.keys(packageJSON.packageDependencies)) {
+    const packagePath = path.join('node_modules', packageName);
```
Comment (aminya) [inline script/generate-metadata-for-builder.js:20]: Please use CONFIG to get the absolute path of `node_modules`

File: script/generate-metadata-for-builder.js [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        if (
+          packageMenuPath.endsWith('.cson') ||
+          packageMenuPath.endsWith('.json')
+        ) {
+          packageNewMetadata.menus[packageMenuPath] = CSON.readFileSync(
+            packageMenuPath
+          );
+        }
+      }
+    }
```
Comment (aminya) [inline script/generate-metadata-for-builder.js:120]: Each of these blocks of code can become a function with a descriptive name and good documentation.

File: package.json [hunk 13]
```diff
@@ -1,5 +1,6 @@
 {
   "name": "atom",
+  "author": "Atom Community <atom@atom.com.br>",
```
Comment (aminya) [inline package.json:3]: The correct email is [atomcommunity@outlook.com](mailto:atomcommunity@outlook.com)

File: package.json [hunk 14]
```diff
@@ -161,7 +161,7 @@
     "temp": "0.9.2",
     "text-buffer": "^13.18.5",
     "timecop": "https://www.atom.io/api/packages/timecop/versions/0.36.2/tarball",
-    "tree-sitter": "git+https://github.com/DeeDeeG/node-tree-sitter.git#bb298eaae66e0c4f11908cb6209f3e141884e88e",
+    "tree-sitter": "^0.17.2",
```
Comment (aminya) [inline package.json:164]: Should be a different pull request

File: package.json [hunk 15]
```diff
@@ -69,7 +70,6 @@
     "fuzzy-finder": "https://www.atom.io/api/packages/fuzzy-finder/versions/1.14.3/tarball",
     "git-diff": "file:packages/git-diff",
     "git-utils": "5.7.1",
-    "github": "https://www.atom.io/api/packages/github/versions/0.36.10/tarball",
```
Comment (aminya) [inline package.json:72]: Why remove GitHub from Atom?

**Review by aminya — CHANGES_REQUESTED** [review]
Add more comments in case we get back to this. But do not spend a lot of time on it for now.

**Comment by mauricioszabo** [comment]
Ok, so this PR was to try an alternative build method that did not depend on scripts at all (or the least we can). The way it's currently moving, and considering priorities, this is probably add another moving piece to what is somewhat too many moving pieces, so I'll close this PR. We may re-organize ideas in the future, when higher priorities are attended.

Also, to handle the new requirements of using electron-builder over the transpiled version of packages and everything, this PR will need to change quite a lot; finally, as we're now, I'm monkey-patching Electron Builder to not remove some metadata that we actually need, so again, I'm not sure this is useful right now.

So, closing for now as to not clutter the PRs we actually need to attend, we may re-visit this in the future.

**Comment by aminya** [comment]
Please keep it open so it is not lost
